08.ls_object/lib/filedata.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require 'etc'
+require 'date'
+
+class FileData
+  FILETYPE_CONVERTING_CODE = {
+    'file' => '-',
+    'directory' => 'd',
+    'link' => 'l'
+  }.freeze
+
+  PERMISSION_CONVERTING_CODE = {
+    0 => '---',
+    1 => '--x',
+    2 => '-w-',
+    3 => '-wx',
+    4 => 'r--',
+    5 => 'r-x',
+    6 => 'rw-',
+    7 => 'rwx'
+  }.freeze
+
+  attr_reader :target, :file, :type, :mode, :nlink, :uid, :gid, :size, :updated_at, :blocks
+
+  def initialize(target, file)
+    @target = target
+    @file = file
+    @file_path = "#{target}/#{file}"
+    fs = File.lstat(@file_path)
+    @type = fs.ftype
+    @mode = fs.mode
+    @nlink = fs.nlink.to_s
+    @uid = Etc.getpwuid(fs.uid).name
+    @gid = Etc.getgrgid(fs.gid).name
+    @size = fs.size.to_s
+    @updated_at = fs.mtime
+    @blocks = fs.blocks
+  end
+
+  def date
+    time_or_year = older_than_month?(6) ? ' %Y' : '%H:%M'
+    updated_at.strftime("%_m %_d #{time_or_year}")
+  end
+
+  def older_than_month?(month)
+    updated_at.to_date <= Date.today.prev_month(month)
+  end
+
+  def filetype_permission
+    "#{filetype}#{permission}"
+  end
+
+  def filetype
+    FILETYPE_CONVERTING_CODE[type]
+  end
+
+  def permission
+    mode.to_s(8)[-3, 3].chars.map do |num|
+      PERMISSION_CONVERTING_CODE[num.to_i]
+    end.join
```

Comment:
`map`なので手続き(`do...end`)よりも式('{...}`のほうがよさそう。
すべて繋げると意図の説明がまったく無いので:

```suggestion
    octal_digits = mode.to_s(8)[-3, 3].chars
    octal_digits.map { |oct| PERMISSION_CONVERTING_CODE[oct.to_i] }.join
```

ぐらいは説明があってもよさそうです。

`PERMISSION_CONVERTING_CODE` が長いかもしれませんね。役割に照らすと、
`PERMISSION_LABELS`ぐらいでいいかもしれません。

```ruby
mode.to_s(8)[-3, 3].chars.map { |oct| PERMISSION_LABELS[oct.to_i] }.join
```

短いけど、読み下すのがしんどそうですね。

```ruby
mode.to_s(8)[-3, 3].chars
  .map { |oct| PERMISSION_LABELS[oct.to_i] }.join
```

ともできますが、こうするなら、`octal_digits`変数を導入した版と大差なさそう(2行)。

定数のキーを、 `{0 => '---',...}`ではなくて、`{ '0' => '---',...}` と明示的に文字列にすれば、`to_i`は不要になりますね。

Reply:
c30ffd4
e20f585